src/selectors/layers.js

```diff
@@ -0,0 +1,51 @@
+import { createSelector } from 'reselect';
+import { getLayoutNodes, getGraphSize } from './layout';
+import { getVisibleLayerIDs } from './disabled';
+
+const getLayerName = state => state.layer.name;
+
+/**
+ * Get layer positions
+ */
+export const getLayers = createSelector(
+  [getLayoutNodes, getVisibleLayerIDs, getLayerName, getGraphSize],
+  (nodes, layerIDs, layerName, { width }) => {
+    // Get list of layer Y positions from nodes
+    const layerY = nodes.reduce((layerY, node) => {
+      if (!layerY[node.layer]) {
+        layerY[node.layer] = [];
+      }
+      layerY[node.layer].push(node.y);
+      return layerY;
+    }, {});
+
+    /**
+     * Determine the y position and height of a layer band
+     * @param {number} id
+     */
+    const calculateYPos = (layerID, prevID, nextID) => {
+      const yMin = Math.min(...layerY[layerID]);
+      const yMax = Math.max(...layerY[layerID]);
+      const prev = layerY[prevID];
+      const next = layerY[nextID];
+      const topYGap = prev && yMin - Math.max(...prev);
+      const bottomYGap = next && Math.min(...next) - yMax;
+      const yGap = (topYGap || bottomYGap) / 2;
+      const y = yMin - yGap;
+      const height = yMax + yGap - y;
+      return { y, height };
+    };
+
+    return layerIDs.map((id, i) => {
+      const prevID = layerIDs[i - 1];
+      const nextID = layerIDs[i + 1];
+      return {
```

Comment:
I originally did it that way. But because there are many nodes and only a few layers, I thought this should be faster. I'll illustrate. Assume layer count = 6 and node count = 200.

Option 1: 
Loop through the list of nodes to get a list of layer Y positions. Then loop through the list of layers to get their height and padding by comparing them to their neighbouring layers.

Option 2: 
Loop through the list of layers. For each layer, loop through the list of nodes to get the y position for this layer and its neighbouring layer, then calculate height and padding.

I'm bad at Big O notation but I'm pretty sure Option 1 should be faster as it requires fewer overall steps, no?

Reply:
You're right, I think I didn't quite catch that it was going over nodes and not layers at the start, so makes more sense now, thanks for clarifying.